runtime/core/decoder/onnx_asr_model.cc

```diff
@@ -0,0 +1,347 @@
+
+    att_mask_.resize(required_cache_size + chunk_size_);
+    for (size_t i = 0; i < required_cache_size + chunk_size_; ++i) {
+      att_mask_[i] = i >= required_cache_size;
+    }
+    const int64_t att_mask_shape[] = {1, 1, required_cache_size + chunk_size_};
+    att_mask_ort_ = Ort::Value::CreateTensor<bool>(
+        memory_info_, reinterpret_cast<bool*>(att_mask_.data()),
+        att_mask_.size(), att_mask_shape, 3);
+  } else {
+    att_cache_.resize(0);
+    const int64_t att_cache_shape[] = {num_blocks_, head_, 0,
+                                       encoder_output_size_ / head_ * 2};
+    att_cache_ort_ = Ort::Value::CreateTensor<float>(
+        memory_info_, att_cache_.data(), att_cache_.size(), att_cache_shape, 4);
+
+    att_mask_.resize(0);
+    const int64_t att_mask_shape[] = {0, 0, 0};
+    att_mask_ort_ = Ort::Value::CreateTensor<bool>(
+        memory_info_, reinterpret_cast<bool*>(att_mask_.data()),
+        att_mask_.size(), att_mask_shape, 3);
+  }
+
+  cnn_cache_.resize(num_blocks_ * encoder_output_size_ *
+                    (cnn_module_kernel_ - 1));
+  const int64_t cnn_cache_shape[] = {num_blocks_, 1, encoder_output_size_,
+                                     cnn_module_kernel_ - 1};
+
+  cnn_cache_ort_ = Ort::Value::CreateTensor<float>(
+      memory_info_, cnn_cache_.data(), cnn_cache_.size(), cnn_cache_shape, 4);
+
+  encoder_outs_.clear();
+}
+
+void OnnxAsrModel::ForwardEncoderFunc(
+    const std::vector<std::vector<float>>& chunk_feats,
+    std::vector<std::vector<float>>* out_prob) {
+  int num_frames = cached_feature_.size() + chunk_feats.size();
+  const int feature_dim = chunk_feats[0].size();
+  int64_t requried_cache_size = chunk_size_ * num_left_chunks_;
+  std::vector<float> model_input;
+  for (size_t i = 0; i < cached_feature_.size(); ++i) {
+    for (int j = 0; j < feature_dim; j++) {
+      model_input.emplace_back(cached_feature_[i][j]);
+    }
+  }
+  for (size_t i = 0; i < chunk_feats.size(); ++i) {
+    for (int j = 0; j < feature_dim; j++) {
+      model_input.emplace_back(chunk_feats[i][j]);
+    }
+  }
+  const int64_t input_shape[3] = {1, num_frames, feature_dim};
+  Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
+      memory_info_, model_input.data(), model_input.size(), input_shape, 3);
+  Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>(
+      memory_info_, &offset_, 1, std::vector<int64_t>{}.data(), 0);
+
+  Ort::Value requried_cache_size_tensor = Ort::Value::CreateTensor<int64_t>(
+      memory_info_, &requried_cache_size, 1, std::vector<int64_t>{}.data(), 0);
+
+  std::vector<Ort::Value> tensors;
+  tensors.emplace_back(std::move(input_tensor));
+  tensors.emplace_back(std::move(offset_tensor));
+  tensors.emplace_back(std::move(requried_cache_size_tensor));
+  tensors.emplace_back(std::move(att_cache_ort_));
+  tensors.emplace_back(std::move(cnn_cache_ort_));
+  tensors.emplace_back(std::move(att_mask_ort_));
```

Comment:
In 16/4 mode, `at_mask_ort_` should be reused for every chunk, however, `move` will make `att_mask_ort_` inaccessible, leading to segment fault. We should recreate `att_mask_ort_` everytime we feed a new chunk.

Reply:
option-2: there is no need to maintain `att_mask_ort_` as a class member, just construct `att_mask_ort` on-the-fly wherever needed.